packages/asset-value/src/assetValue.ts

```diff
@@ -0,0 +1,212 @@
+import { AssetId } from '@shapeshiftoss/caip'
+import { BigNumber } from 'bignumber.js'
+import { Md5 } from 'ts-md5'
+
+import { CHECKSUM_LENGTH, DEFAULT_FORMAT_DECIMALS } from './constants'
+import { AssetValueParams, isAssetValueParams, SerializedAssetValue } from './types'
+import { AssetValueFormat } from './types'
+import { bn, bnOrZero } from './utils'
+
+type AssetValueState = {
+  assetId: AssetId
+  precision: number
+  value: string
+}
+
+//TODO(pastaghost): Add JSDoc documentation
+export class AssetValue {
+  state: AssetValueState
+
+  public constructor(params: AssetValueParams | SerializedAssetValue) {
+    if (isAssetValueParams(params)) {
+      if (params.asset && !(params.asset.assetId?.length && params.asset.precision > -1)) {
+        throw new Error('Cannot initialize AssetValue with invalid asset')
+      }
+
+      if (!params.asset && !(params.assetId?.length && params.precision > -1)) {
+        throw new Error('Cannot initialize AssetValue with invalid asset')
+      }
+
+      let _value: string
+      switch (params.format) {
+        case AssetValueFormat.BASE_UNIT:
+          _value = bnOrZero(params.value).toFixed(0, BigNumber.ROUND_DOWN)
+          break
+        case AssetValueFormat.PRECISION:
+          _value = bnOrZero(params.value)
+            .multipliedBy(bn(10).pow(params.asset ? params.asset.precision : params.precision))
+            .toString()
+          break
+      }
+      this.state = {
+        assetId: params.asset ? params.asset.assetId : params.assetId,
```

Comment:
shorthand option available
```suggestion
        assetId: params.assetId ?? params.asset.assetId,
```